ZCS-11458: Adding new LDAP attributes to let admins schedule backups …

rcyarrapothu opened the pull request:
Feature:
Adding the flexibility where admins can schedule backups from UI not just through CLI.

Implementation:
The eventual plan is to move away from crontab to a java based scheduler like service. As part of this plan, we are adding the LDAP attributes where admins can schedule the backup using them from UI instead of doing it through the zmschedulebackup CLI command as the first step. These attributes will be used by ScheduleBackupsRequest to set the crontab schedule from UI.

Newly added attributes: 

- zimbraGlobalBackupEnabled: TRUE
- zimbraBackupSyncUsersEnabled: FALSE
- zimbraBackupStartTime: 01:00
- zimbraBackupRetentionDays: 8
- zimbraBackupBlobsCompressType: zip
- zimbraBackupSyncEnabled: FALSE
- zimbraMailReportEnabled: TRUE
- zimbraBackupStartDay: 6
- zimbraBackupObjectLevelEnabled: FALSE
- zimbraBackupCrontabConfig: FALSE

Updates:

- zimbraDumpsterEnabled - Added in GlobalConfig, domain and flag for domainInherited
- zimbraMailDumpsterLifetime - Added in GlobalConfig, domain and flag for domainInherited
- zimbraDumpsterPurgeEnabled -  Added in GlobalConfig, domain and flag for domainInherited
- zimbraDumpsterUserVisibleAge - Added in GlobalConfig, domain and flag for domainInherited

Re-use from ZimbraOS:

- zimbraBackupEnabled with ID 9014

Related PRs:

https://github.com/Zimbra/zm-core-utils/pull/118
https://github.com/Zimbra/zm-backup-store/pull/28

rcyarrapothu commented on store/conf/attrs/zimbra-attrs.xml:
```diff
@@ -9925,8 +9925,51 @@ TODO: delete them permanently from here
+  </attr>
+<attr id="4010" name="zimbraBackupStartTime" type="duration" cardinality="single" optionalIn="globalConfig,server"  flags="serverInherited" since="9.1.0">
+    <globalConfigValue>2</globalConfigValue>
+    <desc>Starts backup each day at this time. This attribute is used for cronjob only to set hour</desc>
```
Corrected. This value can take minutes from UI now in HH:mm

k-kato commented on store/conf/attrs/zimbra-attrs.xml:
```diff
@@ -9925,8 +9925,51 @@ TODO: delete them permanently from here
 </attr>
+<attr id="4007" name="zimbraBackupConfigEnabled" type="boolean" cardinality="single" optionalIn="globalConfig,domain,cos,account" flags="accountCosDomainInherited,domainInherited" since="9.1.0">
+    <desc>Whether or not account is eligible for backup</desc>
+  </attr>
+<attr id="4008" name="zimbraBackupEnabled" type="boolean" cardinality="single" optionalIn="globalConfig" since="9.1.0">
```
add an empty line. Others as well.
```suggestion
  </attr>

<attr id="4008" name="zimbraBackupEnabled" type="boolean" cardinality="single" optionalIn="globalConfig" since="9.1.0">
```

rcyarrapothu replied:
Updated.

k-kato commented on store/conf/attrs/zimbra-attrs.xml:
```diff
@@ -9925,8 +9925,51 @@ TODO: delete them permanently from here
   <defaultCOSValue>30</defaultCOSValue>
   <desc>Default calendar resolution preference</desc>
 </attr>
+<attr id="4007" name="zimbraBackupConfigEnabled" type="boolean" cardinality="single" optionalIn="globalConfig,domain,cos,account" flags="accountCosDomainInherited,domainInherited" since="9.1.0">
+    <desc>Whether or not account is eligible for backup</desc>
+  </attr>
```
Fix indent. Others as well.

rcyarrapothu replied:
Updated.

k-kato commented on store/conf/attrs/zimbra-attrs.xml:
```diff
@@ -9925,8 +9925,51 @@ TODO: delete them permanently from here
+  </attr>
+<attr id="4012" name="zimbraBackupBlobsCompress" type="boolean" cardinality="single" optionalIn="globalConfig,server" flags="serverInherited" since="9.1.0">
+    <globalConfigValue>TRUE</globalConfigValue>
+    <desc>If True, Blobs inside backups are compressed in zip format by default unless specified</desc>
```
```suggestion
    <desc>If true, blobs inside backups are compressed in zip format by default unless specified</desc>
```

rcyarrapothu replied:
Updated.

k-kato commented on store/conf/attrs/zimbra-attrs.xml:
```diff
@@ -9925,8 +9925,51 @@ TODO: delete them permanently from here
+  </attr>
+<attr id="4008" name="zimbraBackupEnabled" type="boolean" cardinality="single" optionalIn="globalConfig" since="9.1.0">
+    <globalConfigValue>TRUE</globalConfigValue>
+    <desc>If true, enables Backup globally and if there is pre-configured data the crontab should get that schedule</desc>
```
```suggestion
    <desc>If true, enables backup globally and if there is pre-configured data the crontab should get that schedule</desc>
```

k-kato commented on store/conf/attrs/zimbra-attrs.xml:
```diff
@@ -9925,8 +9925,51 @@ TODO: delete them permanently from here
+    <desc>If True, Blobs inside backups are compressed in zip format by default unless specified</desc>
+  </attr>
+<attr id="4013" name="zimbraBackupBlobsCompressZipStore" type="boolean" cardinality="single" optionalIn="globalConfig,server" flags="serverInherited" since="9.1.0">
+    <globalConfigValue>FALSE</globalConfigValue>
```
Just to confirm. Is the default value correct?

k-kato commented on store/conf/attrs/zimbra-attrs.xml:
```diff
@@ -9925,8 +9925,51 @@ TODO: delete them permanently from here
+  </attr>
+<attr id="4014" name="zimbraBackupSyncMode" type="boolean" cardinality="single" optionalIn="globalConfig,server" flags="serverInherited" since="9.1.0">
+    <globalConfigValue>FALSE</globalConfigValue>
+    <desc>Runs full backup synchronously</desc>
```
It is unclear. Add more detail.

k-kato commented on store/conf/attrs/zimbra-attrs.xml:
```diff
@@ -9925,8 +9925,51 @@ TODO: delete them permanently from here
+    <globalConfigValue>FALSE</globalConfigValue>
+    <desc>Backup up blobs in zip files without compression.</desc>
+  </attr>
+<attr id="4014" name="zimbraBackupSyncMode" type="boolean" cardinality="single" optionalIn="globalConfig,server" flags="serverInherited" since="9.1.0">
```
The attr name is `Mode` but the type is `boolean`. Mismatch.

k-kato commented on store/conf/attrs/zimbra-attrs.xml:
```diff
@@ -9925,8 +9925,51 @@ TODO: delete them permanently from here
+  </attr>
+<attr id="4015" name="zimbraMailReportEnabled" type="boolean" cardinality="single" optionalIn="globalConfig,server" flags="serverInherited" since="9.1.0">
+    <globalConfigValue>TRUE</globalConfigValue>
+    <desc>If true, send an email report to the admin user.</desc>
```
Just to confirm. Is there any attribute to specify destination email address?

rcyarrapothu replied:
Yes, we have this existing attribute: zimbraBackupReportEmailRecipients

k-kato replied:
I got it. Thanks!

k-kato commented on store/conf/attrs/zimbra-attrs.xml:
```diff
@@ -9925,8 +9925,51 @@ TODO: delete them permanently from here
+    <globalConfigValue>TRUE</globalConfigValue>
+    <desc>If true, enables Backup globally and if there is pre-configured data the crontab should get that schedule</desc>
+  </attr>
+<attr id="4009" name="zimbraBackupSyncUsers" type="boolean" cardinality="single" optionalIn="globalConfig" since="9.1.0">
```
Prefer `zimbraBackupSyncUsersEnabled` for boolean

k-kato commented on store/conf/attrs/zimbra-attrs.xml:
```diff
@@ -9925,8 +9925,51 @@ TODO: delete them permanently from here
+    <globalConfigValue>8</globalConfigValue>
+    <desc>The number of days backups will be kept. Backup Scheduler deletes backups older than the set value</desc>
+  </attr>
+<attr id="4012" name="zimbraBackupBlobsCompress" type="boolean" cardinality="single" optionalIn="globalConfig,server" flags="serverInherited" since="9.1.0">
```
Prefer `zimbraBackupBlobsCompressEnabled` for boolean

k-kato commented on store/conf/attrs/zimbra-attrs.xml:
```diff
@@ -9925,8 +9925,51 @@ TODO: delete them permanently from here
+    <globalConfigValue>TRUE</globalConfigValue>
+    <desc>If True, Blobs inside backups are compressed in zip format by default unless specified</desc>
+  </attr>
+<attr id="4013" name="zimbraBackupBlobsCompressZipStore" type="boolean" cardinality="single" optionalIn="globalConfig,server" flags="serverInherited" since="9.1.0">
```
Prefer `zimbraBackupBlobsCompressZipStoreEnabled` for boolean

rcyarrapothu replied:
Removed this attribute.

k-kato commented on store/conf/attrs/zimbra-attrs.xml:
```diff
@@ -9925,8 +9925,51 @@ TODO: delete them permanently from here
+    <globalConfigValue>TRUE</globalConfigValue>
+    <desc>If true, send an email report to the admin user.</desc>
+  </attr>
+<attr id="4016" name="zimbraBackupStandardFullBackup" type="astring" max="256" cardinality="single" optionalIn="globalConfig,server" flags="serverInherited" since="9.1.0">
+    <globalConfigValue>7</globalConfigValue>
+    <desc>Indicates a scheduled backup start day for full-backup</desc>
+  </attr>
```
* Is this mean that backup runs on 7th day in every month? 7 Jan, 7 Feb, ... ? Fix desc to be exact meaning. The attr name may need to be changed.
* Is `astring` appropriate? `integer`?
* `Indicates` may be unnecessary.

rcyarrapothu replied:
Updated.

k-kato commented on store/conf/attrs/zimbra-attrs.xml:
```diff
@@ -9925,8 +9925,51 @@ TODO: delete them permanently from here
+    <globalConfigValue>7</globalConfigValue>
+    <desc>Indicates a scheduled backup start day for full-backup</desc>
+  </attr>
+<attr id="4017" name="ZimbraBackupStandardIncBackup" type="astring" max="256" cardinality="single" optionalIn="globalConfig,server" flags="serverInherited" since="9.1.0">
+    <globalConfigValue>1</globalConfigValue>
+    <desc>Indicates a scheduled backup start day for incremental-backup which by default runs 6 out of 7 days except the day full-backup is triggered.</desc>
+  </attr>
```
* same as zimbraBackupStandardFullBackup. What does `start day` mean? It is not easy to understand how the attr works.
* Is `astring` appropriate? `integer`?
* `Indicates` may be unnecessary.

rcyarrapothu replied:
Removed this attribute. We are now only setting the zimbraBackupStartDay value from the UI which is used on the backend to set the full backup start day and incremental backup schedule for the other 6 days in a week for Standard backup.

k-kato commented on store/conf/attrs/zimbra-attrs.xml:
```diff
@@ -9925,8 +9925,51 @@ TODO: delete them permanently from here
+  </attr>
+<attr id="4012" name="zimbraBackupBlobsCompress" type="boolean" cardinality="single" optionalIn="globalConfig,server" flags="serverInherited" since="9.1.0">
+    <globalConfigValue>TRUE</globalConfigValue>
+    <desc>If True, Blobs inside backups are compressed in zip format by default unless specified</desc>
```
Regarding `compressed in zip`, it is not always "compressed" according to the description of zimbraBackupBlobsCompressZipStore.

rcyarrapothu replied:
Removed this attribute

k-kato commented on store/conf/attrs/zimbra-attrs.xml:
```diff
@@ -9925,8 +9925,51 @@ TODO: delete them permanently from here
+    <globalConfigValue>TRUE</globalConfigValue>
+    <desc>If True, Blobs inside backups are compressed in zip format by default unless specified</desc>
+  </attr>
+<attr id="4013" name="zimbraBackupBlobsCompressZipStore" type="boolean" cardinality="single" optionalIn="globalConfig,server" flags="serverInherited" since="9.1.0">
```
It is not easy to understand the relation/difference between zimbraBackupBlobsCompress and zimbraBackupBlobsCompressZipStore.

zimbraBackupBlobsCompress TRUE and zimbraBackupBlobsCompressZipStore TRUE -> how are blob files taken backup?
TRUE and FALSE -> ?
FALSE and TRUE -> ?
FALSE and FALSE -> ?

rcyarrapothu replied:
zimbraBackupBlobsCompressType is used to get this value now and updated the description

k-kato commented:
just in case, please click "Load more" to see all my comments.

rcyarrapothu commented on store/conf/attrs/zimbra-attrs.xml:
```diff
@@ -9925,8 +9925,51 @@ TODO: delete them permanently from here
+  </attr>
+<attr id="4009" name="zimbraBackupSyncUsers" type="boolean" cardinality="single" optionalIn="globalConfig" since="9.1.0">
+    <globalConfigValue>FALSE</globalConfigValue>
+    <desc>If true,  will ensure the same users have backup and Dumpster enabled</desc>
```
Updated.

k-kato commented on store/conf/attrs/zimbra-attrs.xml:
```diff
@@ -9925,8 +9925,68 @@ TODO: delete them permanently from here
 </attr>
+
+<attr id="4007" name="zimbraBackupConfigEnabled" type="boolean" cardinality="single" optionalIn="globalConfig,domain,cos,account" flags="accountCosDomainInherited,domainInherited" since="9.1.0">
+  <desc>Whether or not account is eligible for backup</desc>
+        If true on cos level then backup accounts for cos
+        If unset on cos level but true on domain level then backup domains account.
+        If unset on cos level, unset on domain level but true on account level then specific account will be backed up.
+        If true on cos level , unset on domain level but zimbraDomainDefaultCOSId is set to cos then accounts for that domain will be backed up.
+        If false is set on any level then that account or group of accounts will not be eligible for backup.
```
* remove the 4th item. Instead, describe about zimbraDomainDefaultCOSId in the 1st item. Please modify it if there is any better description.
* move `</desc>` to the last.
```suggestion
  <desc>Whether or not account is eligible for backup.
        If true on cos level then backup accounts for cos. zimbraDomainDefaultCOSId is considered.
        If unset on cos level but true on domain level then backup domains account.
        If unset on cos level, unset on domain level but true on account level then specific account will be backed up.
        If false is set on any level then that account or group of accounts will not be eligible for backup.
  </desc>
```

k-kato commented on store/conf/attrs/zimbra-attrs.xml:
```diff
@@ -9925,8 +9925,68 @@ TODO: delete them permanently from here
+        zipStore - blobs are backed up into zip files without compression.
+        noZip - blobs are backed up as individual files rather in zip files.
+  </desc>
+
```
remove an empty line

k-kato commented on store/conf/attrs/zimbra-attrs.xml:
```diff
@@ -9925,8 +9925,68 @@ TODO: delete them permanently from here
+        0-7 or names allowed, with both 0 and 7 representing Sunday. Monday to Saturday (1-6)
+  </desc>
+</attr>
+
```
remove an empty line

k-kato commented on store/conf/attrs/zimbra-attrs.xml:
```diff
@@ -9925,8 +9925,68 @@ TODO: delete them permanently from here
+  <desc>If true, will synchronize dumpster enabled accounts with backup eligible accounts from zimbraBackupConfigEnabled list.</desc>
+</attr>
+
+<attr id="4010" name="zimbraBackupStartTime" type="string" max="256" cardinality="single" optionalIn="globalConfig,server" flags="serverInherited" since="9.1.0">
```
Is `max="256"` appropriate? Isn't it too long?

rcyarrapothu replied:
We might be reading a timestamp or something else in next iterations with this attribute. I would leave it for now, unless it's really needed.

k-kato commented on store/conf/attrs/zimbra-attrs.xml:
```diff
@@ -9925,8 +9925,68 @@ TODO: delete them permanently from here
+
+<attr id="4012" name="zimbraBackupBlobsCompressType" type="string" cardinality="single" max="256" optionalIn="globalConfig,server" flags="serverInherited" since="9.1.0">
+  <globalConfigValue>zip</globalConfigValue>
+  <desc>This attribute accepts one of the three values from UI. zip, zipStore and noZip.
```
`from UI` is unnecessary

rcyarrapothu replied:
Updated.

k-kato commented on store/conf/attrs/zimbra-attrs.xml:
```diff
@@ -9925,8 +9925,68 @@ TODO: delete them permanently from here
+  <desc>The number of days backups will be kept. Backup Scheduler deletes backups older than the set value.</desc>
+</attr>
+
+<attr id="4012" name="zimbraBackupBlobsCompressType" type="string" cardinality="single" max="256" optionalIn="globalConfig,server" flags="serverInherited" since="9.1.0">
```
Can we add `value="zip,zipStore,noZip"`?

rcyarrapothu replied:
It's always one of the three values. It's a drop-down or radio button from UI.

k-kato replied:
okay, so why don't we add `value="zip,zipStore,noZip"` so that it can be restricted on CLI as well?
FYI, zimbraAccountStatus is restricted to one of six values.
```
<attr id="2" name="zimbraAccountStatus" type="enum" value="active,maintenance,locked,closed,lockout,pending" callback="AccountStatus" cardinality="single" requiredIn="account" flags="domainAdminModifiable">
```

rcyarrapothu replied:
right, enum seems to be a good idea but from CLI users can still send `--zip --zipStore` together. The combinations `--zip --noZip` or `--zipStore --noZip` wouldn't work as they are exclusive. But when `--zip --zipStore` is given together, zipStore takes precedence over zip. But I can change it  enum as it should be ok for this iteration as it will only used from UI

rcyarrapothu replied:
updated `max="32"`

k-kato replied:
> For 9.1, all these attributes will only be used from UI and only one value will be passed.

All attributes must be designed regardless of being changed from UI or CLI. Even if a value is checked/restricted on UI, we always must assume that there is a possibility that administrator configures the value via CLI.
If the value must be one of the three values, it should be defined as emun.

rcyarrapothu replied:
Updated to enum.

rcyarrapothu commented on store/conf/attrs/zimbra-attrs.xml:
```diff
@@ -9925,8 +9925,68 @@ TODO: delete them permanently from here
+        If unset on cos level but true on domain level then backup domains account.
+        If unset on cos level, unset on domain level but true on account level then specific account will be backed up.
+        If true on cos level , unset on domain level but zimbraDomainDefaultCOSId is set to cos then accounts for that domain will be backed up.
+        If false is set on any level then that account or group of accounts will not be eligible for backup.
```
Updated.

k-kato commented on store/conf/attrs/zimbra-attrs.xml:
```diff
@@ -9925,8 +9925,66 @@ TODO: delete them permanently from here
+  <globalConfigValue>7</globalConfigValue>
+  <desc>The day of the week on which the full backup starts.
+        When a day is select their relative value is set on the crontab.
+        0-7 or names allowed, with both 0 and 7 representing Sunday. Monday to Saturday (1-6)
```
> names allowed

Then type should not be integer but string, correct?

rcyarrapothu replied:
Updated.

k-kato reviewed (approved):
LGTM. Thank you @rcyarrapothu for all the discussions and updates.

ronstra-synacor commented on common/src/java/com/zimbra/common/account/ZAttrProvisioning.java:
```diff
@@ -144,6 +144,24 @@ public static AutoProvMode fromString(String s) throws ServiceException {
+        private String mValue;
+        private BackupBlobsCompressType(String value) { mValue = value; }
+        public String toString() { return mValue; }
+        public static BackupBlobsCompressType fromString(String s) throws ServiceException {
+            for (BackupBlobsCompressType value : values()) {
+                if (value.mValue.equals(s)) return value;
+             }
+             throw ServiceException.INVALID_REQUEST("invalid value: "+s+", valid values: "+ Arrays.asList(values()), null);
+        }
```
Looks like you're just following the pattern of other enums here, but there's already a `toString` and `valueOf` on `Enum` that accomplish (almost) the same behavior.

No action necessary. Just sayin'.

silentsakky commented on store/conf/attrs/zimbra-attrs.xml:
```diff
@@ -9925,8 +9925,66 @@ TODO: delete them permanently from here
+  <desc>If true, will synchronize dumpster enabled accounts with backup eligible accounts from zimbraBackupConfigEnabled list.</desc>
+</attr>
+
+<attr id="4010" name="zimbraBackupStartTime" type="string" max="64" cardinality="single" optionalIn="globalConfig,server" flags="serverInherited" since="9.1.0">
```
From UI perspective we are giving facility to set day and time only to admins for now, but having ldap attributes like day and time would not be scalable in future
like in future what if I want to provide facility to admin to set more fine grain control on this like below
- run full back on monday and thursday every week
- run full backup on alternate weeks on every sunday
- run full backup on monthly basis
Can't we have an ldap attribute where you can specify full cron compatible string, which will be used as it is when updating crontab
Same cron entry can be used also when we remove usage of crontab and convert to service

as I am asking this as a requirement change, I am ok to fork this into separate ticket if that is preferable

silentsakky replied:
@shardool-gore @k-kato @zimsuchitgupta @rcyarrapothu  your thoughts?

rcyarrapothu replied:
I agree with you on adding more flexibility to this attribute but it's a string and we can actually pass in different values not just start day or time only. This could be changed to a timestamp easily in future. The logic where we are handling this attribute should be adjusted accordingly. 
And "zimbraBackupStartTime" fits well even if we are passing a full crontab in my opinion.

k-kato replied:
Here is just an idea.
We can define/reserve another attribute to write full crontab setting.
For example,
```
<attr id="xxx" name="zimbraBackupCompleteDefinition" type="string" cardinality="multi" optionalIn="globalConfig,server" flags="serverInherited" since="9.1.0"> 
```
It is `multi`. It will be able to set full backup configuration of crontab like this.
(This is just an example)
```
If these settings want to be set,
   f    0 1 * * 6
   i    0 1 * * 0-5
   d 1m 0 0 * * *

set multiple values:
   zmprov mcf +zimbraBackupCompleteDefinition 'f    0 1 * * 6'
   zmprov mcf +zimbraBackupCompleteDefinition 'i    0 1 * * 0-5'
   zmprov mcf +zimbraBackupCompleteDefinition 'd 1m 0 0 * * *'
```
Need to modify a logic; If it is set, it will overwrites all other backup-related settings, or other settings will be ignored.
If we also need UI, we can add a `textarea` to write the settings.

rcyarrapothu replied:
Created new attribute "zimbraBackupCrontabConfig". It's a reserve attribute for future enhancements.

silentsakky commented on store/conf/attrs/zimbra-attrs.xml:
```diff
@@ -9925,8 +9925,66 @@ TODO: delete them permanently from here
+
+<attr id="4008" name="zimbraBackupEnabled" type="boolean" cardinality="single" optionalIn="globalConfig" since="9.1.0">
+  <globalConfigValue>TRUE</globalConfigValue>
+  <desc>If true, enables backup globally and if there is pre-configured data the crontab will get that schedule.</desc>
```
description not clear enough, what is pre-configured data?

lavegupta commented on store/conf/attrs/zimbra-attrs.xml:
```diff
@@ -9925,8 +9925,80 @@ TODO: delete them permanently from here
+  </desc>
+</attr>
+
+<attr id="4016" name="zimbraBackupObjectLevelEnabled" type="boolean" cardinality="single" optionalIn="globalConfig,server" flags="serverInherited" since="9.1.0">
+  <globalConfigValue>TRUE</globalConfigValue>
+  <desc>If true, --accounts option will be set to zimbraBackupConfigEnabled to zmschedulebackup command. Enables cos,domain,server,account level backup.
+        If false, --accounts option will be set to all accounts to zmschedulebackup command. Backups all accounts.
+  </desc>
+</attr>
```
I think logic should be reverse, as this option says in UI "**backup all users**".
So, 
**If true**, --accounts option will be set to all accounts to zmschedulebackup command. Backups all accounts.
**If false**, --accounts option will be set to zimbraBackupConfigEnabled to zmschedulebackup command. Enables cos,domain,server,account level backup.

rcyarrapothu replied:
I think it should be ok considering we are actually adding this object level backup as a new functionality and backing up "all users" is just a fall backup value. As mentioned in the other comments, we are considering these attributes to be generic and not specific to the UI.

k-kato replied:
Isn't `useBackupConfig` (not `zimbraBackupConfigEnabled`) added if it is true?

lavegupta commented:
`zimbraDumpsterEnabled` should be available to modify at global level. See https://jira.corp.synacor.com/browse/ZCS-11458?focusedCommentId=1921697&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-1921697

rcyarrapothu commented:
zimbraDumpsterEnabled is now available to modify at global level

k-kato commented:
> `zimbraDumpsterEnabled` should be available to modify at global level. See https://jira.corp.synacor.com/browse/ZCS-11458?focusedCommentId=1921697&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-1921697

How will it be addressed with globalConfig when cos/account level is also set? override? What is related to backup action? What is the purpose of the addition in globalConfig? If it means take backup of data existing in dumpster, do we add another attribute rather than modifying the existing one? 
The existing `zimbraDumpsterEnabled` is already managed/used on many production system and changing the scope will make confusion. I don't think we should add it to global level, but if we must be available at globalConfig level for some reason, it should be added to domain level with domainInherited flag. Admin UI will also need to be updated.

k-kato commented on store/conf/attrs/zimbra-attrs.xml:
```diff
@@ -9925,8 +9925,80 @@ TODO: delete them permanently from here
+</attr>
+
+<attr id="4017" name="zimbraBackupCrontabConfig" type="string" cardinality="multi" optionalIn="globalConfig,server" flags="serverInherited" since="9.1.0">
+  <globalConfigValue>FALSE</globalConfigValue>
```
If type is string, the default value is inappropriate.

k-kato commented on store/conf/attrs/zimbra-attrs.xml:
```diff
@@ -9925,8 +9925,80 @@ TODO: delete them permanently from here
+<attr id="4017" name="zimbraBackupCrontabConfig" type="string" cardinality="multi" optionalIn="globalConfig,server" flags="serverInherited" since="9.1.0">
+  <globalConfigValue>FALSE</globalConfigValue>
+  <desc>Configures crontab schedule. Full, incremental and delete cron schedules can be set using this attribute
+        f '0 1 * * 6' 'i 0 1 * * 0-5' d 1m '0 0 * * *' values can be passed
```
Does this mean set multiple settings (`f, i, d` in this example) as a single value? 
What "multiple" values can/should be set? Set `f xxx`, `i xxx` and `d xxx` separately?
Please describe it more user friendly.

k-kato commented on store/conf/attrs/zimbra-attrs.xml:
```diff
@@ -5284,7 +5284,7 @@ TODO: delete them permanently from here
 </attr>
 
-<attr id="1133" name="zimbraMailDumpsterLifetime" type="duration" cardinality="single" optionalIn="account,cos" flags="accountInherited,domainAdminModifiable,accountInfo" since="7.0.0">
+<attr id="1133" name="zimbraMailDumpsterLifetime" type="duration" cardinality="single" optionalIn="globalConfig,account,cos" flags="accountInherited,domainAdminModifiable,accountInfo" since="7.0.0">
```
add domain and domainInherited
```suggestion
<attr id="1133" name="zimbraMailDumpsterLifetime" type="duration" cardinality="single" optionalIn="globalConfig,domain,account,cos" flags="accountInherited,domainInherited,domainAdminModifiable,accountInfo" since="7.0.0">
```

k-kato commented on store/conf/attrs/zimbra-attrs.xml:
```diff
@@ -6601,12 +6601,12 @@ TODO: delete them permanently from here
 </attr>
 
-<attr id="1314" name="zimbraDumpsterUserVisibleAge" type="duration" cardinality="single" optionalIn="account,cos" flags="accountInherited,domainAdminModifiable" since="8.0.0">
+<attr id="1314" name="zimbraDumpsterUserVisibleAge" type="duration" cardinality="single" optionalIn="globalConfig,account,cos" flags="accountInherited,domainAdminModifiable" since="8.0.0">
```
same

k-kato commented on store/conf/attrs/zimbra-attrs.xml:
```diff
@@ -6601,12 +6601,12 @@ TODO: delete them permanently from here
 </attr>
 
-<attr id="1315" name="zimbraDumpsterPurgeEnabled" type="boolean" cardinality="single" optionalIn="account,cos" flags="accountInherited,domainAdminModifiable" since="8.0.0">
+<attr id="1315" name="zimbraDumpsterPurgeEnabled" type="boolean" cardinality="single" optionalIn="globalConfig,account,cos" flags="accountInherited,domainAdminModifiable" since="8.0.0">
```
same

rcyarrapothu replied:
Updated

k-kato commented:
Please do not squash so that we can check what is updated during code review.

rcyarrapothu commented:
> How will it be addressed with globalConfig when cos/account level is also set? override? What is related to backup action? What is the purpose of the addition in globalConfig? If it means take backup of data existing in dumpster, do we add another attribute rather than modifying the existing one?

This has been discussed and addressed. After reviewing it with team, added in GlobalConfig, domain and flag for domainInherited.
